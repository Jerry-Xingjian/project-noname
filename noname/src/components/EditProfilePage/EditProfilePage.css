.main-content {
	width: 60%;
	position: absolute;
	top: 15%;
	left: 15%;
}

.input {
  width: 100%;
  padding: 12px;
  border: 1px solid #ccc;
  border-radius: 4px;
  resize: vertical;
}

#cancel {
	position: absolute;
	left: 75%;
	width: 90px;
	border-radius: 4px;
	color: #cf250D;
	border: solid 2px transparent;
	background: linear-gradient(rgba(255, 255, 255, 0), rgba(255, 255, 255, 0)), linear-gradient(61deg, #cf250D, #ffae00);
	background-origin: border-box;
	background-clip: content-box, border-box;
	box-shadow: 2px 1000px 1px #fff inset;

}

#save {
	position: absolute;
	left: 89%;
	width: 90px;
	border-radius: 4px;
	color: #cf250D;
	border: solid 2px transparent;
	background: linear-gradient(rgba(255, 255, 255, 0), rgba(255, 255, 255, 0)), linear-gradient(61deg, #cf250D, #ffae00);
	background-origin: border-box;
	background-clip: content-box, border-box;
	box-shadow: 2px 1000px 1px #fff inset;

}

#change-picture {
	position: absolute;
	left:15%;
	background: none;
	box-shadow: none;
}


#cancel:hover{
	box-shadow: none;
  color: white;
}
#save:hover{
	box-shadow: none;
  color: white;
}
