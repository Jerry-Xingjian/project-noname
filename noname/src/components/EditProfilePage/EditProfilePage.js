import { React, useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import './EditProfilePage.css';
import Modal from 'react-bootstrap/Modal';
import { Alert, Container } from 'react-bootstrap';
import ImageUploader from '../ImageUploader/ImageUploader';
import { getCurrentUserProfile, updateUserProfile } from '../../utils/api/user';
import uploadFile from '../../utils/api/aws';
import { authenticate } from '../../utils/auth';

function EditProfilePage(props) {
  const { userId } = props;
  const [userProfileData, setUserProfileData] = useState({});
  const [showEditPicture, setShowEditPicture] = useState(false);
  const [newImage, setNewImage] = useState('');
  const [tempPic, setTempPic] = useState('');
  const navigate = useNavigate();
  const [authFlag, setAuthFlag] = useState(true);
  const [showError, setShowError] = useState(true);

  setInterval(() => {
    if (authenticate() === false) {
      setAuthFlag(false);
    }
  }, 1000);

  // Function to hide the error message
  const hideError = () => {
    setShowError(false);
  };

  async function fetchProfileData() {
    const rawUserData = await getCurrentUserProfile();
    setUserProfileData(rawUserData.data.data);
  }

  const fileToDataUrl = (file) => new Promise((resolve) => {
    const reader = new FileReader();
    reader.onload = (event) => {
      resolve(event.target.result);
    };
    reader.readAsDataURL(file);
  });

  const handleImageUpload = (images) => {
    setNewImage(uploadFile(images[0].file).url);
    fileToDataUrl(images[0].file)
      .then((dataUri) => {
        setTempPic(dataUri);
      });
    setShowEditPicture(false);
  };

  useEffect(() => {
    fetchProfileData();
  }, []);

  async function handleEditProfile(e) {
    e.preventDefault();
    const username = document.getElementById('username').value;
    const email = document.getElementById('email').value;
    const bio = document.getElementById('bio').value;
    let profilePicture;
    if (newImage === '') {
      profilePicture = userProfileData.profilePicture;
    } else {
      profilePicture = newImage;
    }
    // console.log(profilePicture);
    const profile = {
      username, bio, email, profilePicture,
    };
    await updateUserProfile(userId, profile);
    navigate(`/profile/${userId}`);
    // console.log('profile updated');
  }

  let pic;
  if (tempPic === '') {
    pic = userProfileData.profilePicture;
  } else {
    pic = tempPic;
  }

  if (authFlag === false) {
    return (
      <Container
        className="d-flex align-items-center justify-content-center text-center min-vh-100"
      >
        {showError
        && (
        <Alert
          variant="danger"
          onClose={hideError}
          dismissible
        >
          Your session has expired. Please
          {' '}

          <Alert.Link href="/login"> login </Alert.Link>
          {' '}
          again.
        </Alert>
        )}
      </Container>
    );
  }

  return (
    <div className="container">
      <div className="main-content">
        <div className="row">
          <div className="col">
            <button id="change-picture" className="btn btn-link" title="Change Picture" type="button" onClick={() => setShowEditPicture(true)}>
              <img src={pic} className="rounded-circle" alt="Profile" width="175" height="175" />
            </button>
          </div>
          <div className="col">
            <form action={`/profile/${userId}`} onSubmit={handleEditProfile}>
              Username
              <input
                type="text"
                placeholder="Username"
                name="username"
                id="username"
                className="input my-3"
                defaultValue={userProfileData.username}
              />
              <br />
              Bio
              <input
                type="text"
                placeholder="Change your Bio here"
                name="bio"
                id="bio"
                className="input my-3"
                defaultValue={userProfileData.bio}
              />
              <br />
              Email Address
              <input
                type="text"
                placeholder="Email"
                name="email"
                id="email"
                className="input my-3"
                defaultValue={userProfileData.email}
              />
              <br />
              Confirm Password
              <input
                type="password"
                placeholder="*********"
                name="password"
                id="password"
                className="input my-3"
              />
              <br />
              <a type="button" className="btn mx-2 my-2" id="cancel" href={`/profile/${userId}`}>CANCEL</a>
              <input type="submit" value="SAVE" className="btn mx-2 my-2" id="save" />
            </form>
          </div>
        </div>
      </div>
      <Modal
        show={showEditPicture}
        onHide={() => setShowEditPicture(false)}
        className="modal"
      >
        <ImageUploader
          maxNumber={1}
          acceptType={['jpeg', 'jpg', 'png', 'gif', 'tiff']}
          onImagesUpload={handleImageUpload}
        />
      </Modal>
    </div>
  );
}

export default EditProfilePage;
