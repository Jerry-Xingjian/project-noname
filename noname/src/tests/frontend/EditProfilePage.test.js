import React from "react";
import '@testing-library/jest-dom'
import EditProfilePage from "../../components/EditProfilePage/EditProfilePage";
import renderer from 'react-test-renderer'

describe("Snapshot test", () => {
  test("Post card structure test", async () => {
    const component = renderer.create(
      <EditProfilePage />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

