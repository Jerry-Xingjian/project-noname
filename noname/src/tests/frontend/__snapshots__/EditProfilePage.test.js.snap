// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test Post card structure test 1`] = `
<div
  className="container"
>
  <div
    className="main-content"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <button
          className="btn btn-link"
          id="change-picture"
          title="Change Picture"
          type="button"
        >
          <img
            alt="Profile"
            className="rounded-circle"
            height="175"
            src="https://nonome-project-media.s3.amazonaws.com/profile-picture.jpeg"
            width="175"
          />
        </button>
      </div>
      <div
        className="col"
      >
        <form
          action="/profile/1"
        >
          Name
          <input
            className="input my-3"
            id="name"
            name="name"
            placeholder="Username"
            type="text"
          />
          <br />
          Bio
          <input
            className="input my-3"
            id="bio"
            name="bio"
            placeholder="Change your Bio here"
            type="text"
          />
          <br />
          Default Location
          <input
            className="input my-3"
            id="location"
            name="location"
            placeholder="Location"
            type="text"
          />
          <br />
          Password
          <input
            className="input my-3"
            id="password"
            name="password"
            placeholder="*********"
            type="text"
          />
          <br />
          <a
            className="btn mx-2 my-2"
            href="/profile/1"
            id="cancel"
            type="button"
          >
            CANCEL
          </a>
          <input
            className="btn mx-2 my-2"
            id="save"
            type="submit"
            value="SAVE"
          />
        </form>
      </div>
    </div>
  </div>
</div>
`;
